test/e2e/clusterctl_upgrade.go

```diff
@@ -255,16 +273,48 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			input.PreInit(managementClusterProxy)
 		}
 
+		var (
+			coreProvider              string
+			bootstrapProviders        []string
+			controlPlaneProviders     []string
+			infrastructureProviders   []string
+			ipamProviders             []string
+			runtimeExtensionProviders []string
+		)
+
+		coreProvider = input.E2EConfig.GetProviderLatestVersionsByContract(initContract, config.ClusterAPIProviderName)[0]
+		if input.InitWithCoreProvider != "" {
+			coreProvider = input.InitWithCoreProvider
+		}
+
+		getValueOrFallback := func(value []string, fallback []string) []string {
```

Comment:
Let's move this util func outside of the current func ~ down to below `downloadToTmpFile`